File: python/bubble.py
from utility import swap

def bubble_sort(data_set):
    for i in range(len(data_set)):
        change = False

        for j in range(len(data_set) - 1, i, -1):
            if(data_set[j] < data_set[j-1]):
                swap(data_set, j, j-1)
                change = True

        if(not change):
            break

File: go/bubble.go
package main

func bubbleSort(data []float64) {
	for i := len(data) - 1; i > 0; i-- {
		lastSwap := i
		noSwap := true

		for j := 0; j < i; j++ {
			if data[j] > data[j+1] {
				swap(data, j, j+1)

				if noSwap {
					lastSwap = j
				}
				noSwap = false
			}
		}

		if noSwap && lastSwap != i {
			i = lastSwap
		}
	}
}
